Implement WebTransport datagrams and outgoing streams network interface

AnkshitJain opened the pull request:
#### 18fc2c8a829238a023b22f9c584ce09756d4b757
<pre>
Implement WebTransport datagrams and outgoing streams network interface
<a href="https://bugs.webkit.org/show_bug.cgi?id=284069">https://bugs.webkit.org/show_bug.cgi?id=284069</a>
<a href="https://rdar.apple.com/136262968">rdar://136262968</a>

Reviewed by Alex Christensen.

WebTransport streams are hooked up with nw_connection_t for outgoing bidirectional streams, unidirectional streams, and datagrams.
They can now send and receive data.

One test is added, WebTransport.Datagram, and the existing WebTransport.Basic test is now WebTransport.Bidirectional to validate the added changes.
A test for unidirectional stream will be added with the handling of incoming streams.

* Source/WebCore/PAL/pal/spi/cf/CFNetworkSPI.h:
* Source/WebKit/NetworkProcess/webtransport/NetworkTransportSession.cpp:
(WebKit::NetworkTransportSession::sendStreamSendBytes):
(WebKit::NetworkTransportSession::streamSendBytes):
(WebKit::NetworkTransportSession::destroyOutgoingUnidirectionalStream):
(WebKit::NetworkTransportSession::destroyBidirectionalStream):
* Source/WebKit/NetworkProcess/webtransport/NetworkTransportSession.h:
* Source/WebKit/NetworkProcess/webtransport/cocoa/NetworkTransportSessionCocoa.mm:
(WebKit::NetworkTransportSession::NetworkTransportSession):
(WebKit::createDatagramParameters):
(WebKit::NetworkTransportSession::initialize):
(WebKit::NetworkTransportSession::sendDatagram):
(WebKit::NetworkTransportSession::createBidirectionalStream):
(WebKit::NetworkTransportSession::createOutgoingUnidirectionalStream):
(WebKit::NetworkTransportSession::setupListenerConnectionHandler):
(WebKit::NetworkTransportSession::createParameters):
(WebKit::NetworkTransportSession::receiveDatagramLoop):
(WebKit::NetworkTransportSession::createStream):
* Source/WebKit/WebProcess/Network/WebTransportSession.cpp:
(WebKit::WebTransportSession::receiveDatagram):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WebTransport.mm:
(TestWebKitAPI::TEST(WebTransport, ClientBidirectional)):
(TestWebKitAPI::TEST(WebTransport, Datagram)):
(TestWebKitAPI::TEST(WebTransport, Basic)): Deleted.
* Tools/TestWebKitAPI/WebTransportServer.mm:
(TestWebKitAPI::WebTransportServer::WebTransportServer):

Canonical link: <a href="https://commits.webkit.org/287714@main">https://commits.webkit.org/287714@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0f3197f6261171b093a60975448baa510a800e8f

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/80573 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/59580 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/34506 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/85100 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/31555 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/82684 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/68641 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/7885 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/62947 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/20749 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/83642 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/53057 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/73345 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/43252 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/50352 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/27503 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/30014 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/71498 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/28026 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/86528 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/7799 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/5503 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/71235 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/7974 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/69182 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/70476 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/14486 "Passed tests") | [⏳ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/macOS-Ventura-Release-WK2-Intel-Tests-EWS "Waiting to run tests") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/12481 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/7761 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/13280 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/7600 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/11119 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/9405 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebKit/NetworkProcess/webtransport/cocoa/NetworkTransportSessionCocoa.mm:
```diff
@@ -37,121 +37,205 @@
+        return nil;
+    }
+    nw_protocol_stack_t default_stack = nw_parameters_copy_default_protocol_stack(parameters.get());
+    nw_protocol_stack_iterate_application_protocols(default_stack, [&streamType](nw_protocol_options_t options) {
```
streamType needs to be captured by value instead of by reference.  By the time the lambda is called, the stack of NetworkTransportSession::createParameters is probably gone.

achristensen07 commented on Source/WebKit/NetworkProcess/webtransport/cocoa/NetworkTransportSessionCocoa.mm:
```diff
@@ -37,121 +37,205 @@
+RetainPtr<nw_parameters_t> NetworkTransportSession::createParameters(NetworkTransportStreamType streamType)
+{
+    ASSERT(streamType != NetworkTransportStreamType::IncomingUnidirectional);
+    RetainPtr<nw_parameters_t> parameters = adoptNS(nw_parameters_copy(nw_connection_copy_parameters(m_connections.first().get())));
```
&lt;nw_parameters_t&gt; is unnecessary.

achristensen07 commented on Source/WebKit/NetworkProcess/webtransport/cocoa/NetworkTransportSessionCocoa.mm:
```diff
@@ -37,121 +37,205 @@
-        return completionHandler(nullptr);
-
-    RetainPtr endpoint = adoptNS(nw_endpoint_create_host(url.host().utf8().data(), makeString(*port).utf8().data()));
+    RetainPtr<nw_endpoint_t> endpoint = adoptNS(nw_endpoint_create_url(url.string().utf8().data()));
```
&lt;nw_endpoint_t&gt; isn't necessary because of the deduction guides of RetainPtr.  Ditto many times in this PR.

achristensen07 commented on Source/WebKit/NetworkProcess/webtransport/cocoa/NetworkTransportSessionCocoa.mm:
```diff
@@ -37,121 +37,205 @@
 
-    RetainPtr connectionGroup = adoptNS(nw_connection_group_create(descriptor.get(), parameters.get()));
-    if (!connectionGroup) {
+    RetainPtr<nw_listener_t> listener = adoptNS(nw_listener_create_with_connection(connection.get(), parameters.get()));
```
If we want to use nw_listener_create_with_connection it needs to be forward declared in the SPI header.  Can we use API instead?

AnkshitJain replied:
Updated to connection group.

achristensen07 commented on Source/WebKit/NetworkProcess/webtransport/cocoa/NetworkTransportSessionCocoa.mm:
```diff
@@ -37,121 +37,205 @@
+        ASSERT_NOT_REACHED();
+        return nil;
+    }
+    nw_protocol_stack_t default_stack = nw_parameters_copy_default_protocol_stack(parameters.get());
```
I think this needs adoptNS because nw_parameters_copy_default_protocol_stack  has NW_RETURNS_RETAINED

achristensen07 commented on Source/WebKit/NetworkProcess/webtransport/cocoa/NetworkTransportSessionCocoa.mm:
```diff
@@ -37,121 +37,205 @@
+
+void NetworkTransportSession::receiveDatagramLoop()
+{
+    nw_connection_receive(m_connections.first().get(), 1, std::numeric_limits<uint32_t>::max(), makeBlockPtr([weakThis = WeakPtr { *this }] (dispatch_data_t content, nw_content_context_t, bool withFin, nw_error_t error) {
```
We didn't check that m_connections has more than 0 connections in it, so we could be going out of bounds.  If the first connection is special, it should have its own member variable, like m_datagramConnection or something like that.

AnkshitJain replied:
Added a new datagram connection member.

achristensen07 commented on Source/WebKit/NetworkProcess/webtransport/cocoa/NetworkTransportSessionCocoa.mm:
```diff
@@ -37,121 +37,205 @@
+        return completionHandler(std::nullopt);
+    }
+
+    RetainPtr connection = nw_connection_create_with_connection(m_connections.first().get(), parameters.get());
```
This needs adoptNS, and forward declaration in the SPI header.  Is there no API for this?

AnkshitJain replied:
Replaced with connection group.

achristensen07 commented on Tools/TestWebKitAPI/WebTransportServer.mm:
```diff
@@ -40,42 +40,56 @@
 
     Function<Task(Connection)> connectionHandler;
     RetainPtr<nw_listener_t> listener;
-    Vector<RetainPtr<nw_connection_group_t>> connectionGroups;
```
I really don't think moving away from nw_connection_group_t is a good idea.  Can we not just make nw_connection_group_t work in Network.framework instead?

achristensen07 commented on Source/WebKit/NetworkProcess/webtransport/cocoa/NetworkTransportSessionCocoa.mm:
```diff
@@ -37,121 +37,205 @@
+    };
+
+    auto configureQUIC = [](nw_protocol_options_t options) {
+        nw_quic_set_initial_max_streams_bidirectional(options, 1000);
```
These numbers seem arbitrary.  Are they special?  Can they be given a name instead of just using magic numbers?

achristensen07 commented on Source/WebKit/NetworkProcess/webtransport/cocoa/NetworkTransportSessionCocoa.mm:
```diff
@@ -37,121 +37,205 @@
+    auto configureQUIC = [](nw_protocol_options_t options) {
+        nw_quic_set_initial_max_streams_bidirectional(options, 1000);
+        nw_quic_set_initial_max_streams_unidirectional(options, 1000);
+        nw_quic_set_max_datagram_frame_size(options, 65535);
```
Is there a WebTransport spec that says to do this?  Can we use std::numeric_limits<uint16_t>::max() instead?

achristensen07 commented on Source/WebCore/PAL/pal/spi/cocoa/NetworkSPI.h:
```diff
@@ -58,6 +58,18 @@ nw_interface_t nw_path_copy_interface(nw_path_t);
+nw_parameters_t nw_parameters_create_webtransport_http(nw_parameters_configure_protocol_block_t, nw_parameters_configure_protocol_block_t, nw_parameters_configure_protocol_block_t, nw_parameters_configure_protocol_block_t);
+nw_protocol_options_t
+nw_webtransport_create_options(void);
+bool
```
Let's put return values on the same line as the function so each function is one line.

achristensen07 commented on Source/WebKit/NetworkProcess/webtransport/NetworkTransportSession.h:
```diff
@@ -83,15 +85,17 @@ class NetworkTransportSession : public RefCounted<NetworkTransportSession>, publ
 #if PLATFORM(COCOA)
     const RetainPtr<nw_connection_group_t> m_connectionGroup;
     const RetainPtr<nw_endpoint_t> m_endpoint;
+    RetainPtr<nw_connection_t> m_datagramConnection;
```
This should be const, too.  Can we hold off allocating the NetworkTransportSession until we know we have a datagram connection AND a connection group, then pass both to the constructor?

AnkshitJain replied:
This is not feasible as datagramConnection cannot be created before the connection group is started with an incoming connection handler being setup.

achristensen07 commented on Tools/TestWebKitAPI/WebTransportServer.mm:
```diff
@@ -40,42 +40,73 @@
 
-    RetainPtr parameters = adoptNS(nw_parameters_create_quic_stream(NW_PARAMETERS_DEFAULT_CONFIGURATION, configureTLS));
+    auto configureQUIC = [](nw_protocol_options_t options) {
+        nw_quic_set_initial_max_streams_bidirectional(options, 1000);
```
Do we need this?  Can we give 1000 a name?  Can we use std::numeric_limits like we did in the client?

achristensen07 commented on Source/WebKit/NetworkProcess/webtransport/cocoa/NetworkTransportSessionCocoa.mm:
```diff
@@ -141,17 +130,101 @@
+void NetworkTransportSession::createStream(NetworkTransportStreamType streamType, CompletionHandler<void(std::optional<WebTransportStreamIdentifier>)>&& completionHandler)
+{
+    ASSERT(streamType != NetworkTransportStreamType::IncomingUnidirectional);
+#if HAVE(WEB_TRANSPORT)
```
Could we have the entirety of this function in HAVE(WEB_TRANSPORT)?

#if HAVE(WEB_TRANSPORT)
...
#else
completionHandler(std::nullopt);
#endif

Same in NetworkTransportSession::setupDatagramConnection.  We might want to have more code inside WEB_TRANSPORT checks, and just stubs outside.  That may also require running "run-webkit-tests imported/w3c/web-platform-tests/webtransport --reset --no-build" on a system without WEB_TRANSPORT to keep the test bots green, and we may need internal test expectations for these tests.

achristensen07 commented on Source/WebKit/NetworkProcess/webtransport/cocoa/NetworkTransportSessionCocoa.mm:
```diff
@@ -42,116 +42,192 @@
     }
 
-    RetainPtr connectionGroup = adoptNS(nw_connection_group_create(descriptor.get(), parameters.get()));
+    RetainPtr connectionGroup = adoptNS(nw_connection_group_create(nw_group_descriptor_create_multiplex(endpoint.get()), parameters.get()));
```
nw_group_descriptor_create_multiplex needs adoptNS or it will be leaked memory.

achristensen07 commented on Source/WebKit/NetworkProcess/webtransport/cocoa/NetworkTransportSessionCocoa.mm:
```diff
@@ -42,116 +42,192 @@
+    }
+
+    nw_connection_set_state_changed_handler(m_datagramConnection.get(), makeBlockPtr([
+        refThis = RefPtr { this },
```
WebKit style says to use "protectedThis" for this capture name.
Also, it can be a Ref and you don't need to null check it below.

achristensen07 commented on Tools/TestWebKitAPI/WebTransportServer.mm:
```diff
@@ -40,42 +40,73 @@
 WebTransportServer::WebTransportServer(Function<Task(Connection)>&& connectionHandler)
     : m_data(Data::create(WTFMove(connectionHandler)))
 {
+    RetainPtr<nw_parameters_t> parameters = nullptr;
```
= nullptr is not necessary because RetainPtr has a default constructor that does that for you.

achristensen07 commented on Tools/TestWebKitAPI/WebTransportServer.mm:
```diff
@@ -40,42 +40,73 @@
+    RetainPtr<nw_parameters_t> parameters = nullptr;
+    RetainPtr<nw_protocol_options_t> webtransportOptions = nullptr;
+
+#if HAVE(WEB_TRANSPORT)
```
We could probably just put the entirety of WebTransportServer.h and .mm inside HAVE(WEB_TRANSPORT).

achristensen07 commented on Source/WebKit/NetworkProcess/webtransport/cocoa/NetworkTransportSessionCocoa.mm:
```diff
@@ -42,116 +42,192 @@
+        ASSERT_NOT_REACHED();
+        return completionHandler(std::nullopt);
+    }
+    nw_webtransport_options_set_is_unidirectional(webtransportOptions.get(), streamType == NetworkTransportStreamType::OutgoingUnidirectional);
```
What about incoming unidirectional?

achristensen07 commented on Source/WebKit/NetworkProcess/webtransport/cocoa/NetworkTransportSessionCocoa.mm:
```diff
@@ -42,116 +42,205 @@
+        switch (state) {
+        case nw_connection_state_waiting:
+        case nw_connection_state_preparing: {
+            protectedThis = nullptr;
```
I think this line should be removed.

Commit:
Implement WebTransport datagrams and outgoing streams network interface
https://bugs.webkit.org/show_bug.cgi?id=284069
rdar://136262968

Reviewed by Alex Christensen.

WebTransport streams are hooked up with nw_connection_t for outgoing bidirectional streams, unidirectional streams, and datagrams.
They can now send and receive data.

One test is added, WebTransport.Datagram, and the existing WebTransport.Basic test is now WebTransport.Bidirectional to validate the added changes.
A test for unidirectional stream will be added with the handling of incoming streams.

* Source/WebCore/PAL/pal/spi/cf/CFNetworkSPI.h:
* Source/WebKit/NetworkProcess/webtransport/NetworkTransportSession.cpp:
(WebKit::NetworkTransportSession::sendStreamSendBytes):
(WebKit::NetworkTransportSession::streamSendBytes):
(WebKit::NetworkTransportSession::destroyOutgoingUnidirectionalStream):
(WebKit::NetworkTransportSession::destroyBidirectionalStream):
* Source/WebKit/NetworkProcess/webtransport/NetworkTransportSession.h:
* Source/WebKit/NetworkProcess/webtransport/cocoa/NetworkTransportSessionCocoa.mm:
(WebKit::NetworkTransportSession::NetworkTransportSession):
(WebKit::createDatagramParameters):
(WebKit::NetworkTransportSession::initialize):
(WebKit::NetworkTransportSession::sendDatagram):
(WebKit::NetworkTransportSession::createBidirectionalStream):
(WebKit::NetworkTransportSession::createOutgoingUnidirectionalStream):
(WebKit::NetworkTransportSession::setupListenerConnectionHandler):
(WebKit::NetworkTransportSession::createParameters):
(WebKit::NetworkTransportSession::receiveDatagramLoop):
(WebKit::NetworkTransportSession::createStream):
* Source/WebKit/WebProcess/Network/WebTransportSession.cpp:
(WebKit::WebTransportSession::receiveDatagram):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WebTransport.mm:
(TestWebKitAPI::TEST(WebTransport, ClientBidirectional)):
(TestWebKitAPI::TEST(WebTransport, Datagram)):
(TestWebKitAPI::TEST(WebTransport, Basic)): Deleted.
* Tools/TestWebKitAPI/WebTransportServer.mm:
(TestWebKitAPI::WebTransportServer::WebTransportServer):

Canonical link: https://commits.webkit.org/287714@main
